enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1966,10 +1987,169 @@ func newPartitionEvictor(part disk.Partition, fileStorer filestore.Store, blobDi
 	pe.acCount = acCount
 	pe.lru.UpdateSizeBytes(sizeBytes)
 
+	// Sample some random groups at startup so that eviction has something to
+	// work with.
+	// AC eviction requires version 2 or higher.
+	if vg.minDatabaseVersion() >= filestore.Version2 && *groupIDSamplingEnabled {
+		db, err := dbg.DB()
+		if err != nil {
+			return nil, err
+		}
+		defer db.Close()
+
+		approxTotalCount := int64(0)
+		seenGroupIDs := make(map[string]struct{})
+		var approxCounts []groupIDApproxCount
+		for i := 0; i < *groupIDSamplesOnStartup; i++ {
+			groupID, err := pe.sampleGroupID()
+			if status.IsNotFoundError(err) {
+				continue
+			}
+			if err != nil {
+				log.Warningf("Could not sample group ID: %s", err)
+				continue
+			}
+			if _, ok := seenGroupIDs[groupID]; ok {
+				continue
+			}
+			seenGroupIDs[groupID] = struct{}{}
+
+			approxCount, err := pe.approxGroupItemCount(groupID)
+			if err != nil {
+				log.Warningf("Could not approximate count for group ID %s: %s", groupID, err)
+				continue
+			}
+			log.Debugf("Sampled group ID %q with approx item count %d", groupID, approxCount)
+			approxTotalCount += approxCount
+			approxCounts = append(approxCounts, groupIDApproxCount{groupID: groupID, count: approxCount})
+		}
+
+		slices.SortFunc(approxCounts, func(a, b groupIDApproxCount) bool {
+			return a.count < b.count
+		})
+
+		cumulativeCount := int64(0)
+		for i, v := range approxCounts {
+			cumulativeCount += v.count
+			approxCounts[i].cumulativeCount = cumulativeCount
+		}
+
+		pe.groupIDApproxTotalCount = approxTotalCount
+		pe.groupIDApproxCounts = approxCounts
+	}
+
```

Comment:
is it not safe to call updateGroupIDApproxCounts here yet because pe is not fully initialized? Wondering if there's a way to dedupe this code by just calling the real method 1) on startup if necessary and 2) from our periodic timer

Reply:
I was just trying to avoid doing unnecessary work since updateGroupIDApproxCounts does some things that are not necessary when we are populating the list for the first time, but i guess it doesn't really matter so I simplified it.